internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,200 @@
+
+	// If true, do not swap the context to the background context when running.
+	// By default (when this is false), the context will be swapped to the background context when !WaitForCompletion.
+	// The swap is so that jobs do not get canceled when an HTTP request is complete.
+	// Setting this true is useful for testing (as one round of job tests can be wrapped up).
+	RetainOriginalContext bool `json:"-"`
+
+	PoolSize int `json:"-"`
+
+	LockKey string `json:"-"`
+}
+
+func (this *JobOptions) Validate() error {
+	if this == nil {
+		return fmt.Errorf("Job options are nil.")
+	}
+
+	if this.LockKey == "" {
+		return fmt.Errorf("Cannot have an empty lock key.")
+	}
+
+	if this.PoolSize <= 0 {
+		return fmt.Errorf("Pool size must be positive, got %d.", this.PoolSize)
+	}
+
+	if this.Context == nil {
+		this.Context = context.Background()
+	}
+
+	if !this.WaitForCompletion && !this.RetainOriginalContext {
+		this.Context = context.Background()
+	}
+
+	return nil
+}
+
+// RunJob() executes a potentially long-running job over a list of input items.
+// It supports optional caching, cache removal, synchronization, and context cancellation.
+// Given job options, input items, an optional cache lookup/removal function, and a work function,
+// RunJob() processes each item in a parallel pool.
+// Returns the result list, number of remaining items, total run time, and an error.
+// If the context is canceled, returns nil, 0, 0, nil.
+func RunJob[InputType any, OutputType any](options *JobOptions, workItems []InputType, cacheFunc func([]InputType) ([]OutputType, []InputType, error), removeCacheFunc func([]InputType) error, workFunc func(InputType) (OutputType, int64, error)) ([]OutputType, int, int64, error) {
+	if workFunc == nil {
+		return nil, len(workItems), 0, fmt.Errorf("Cannot run job with a nil work function.")
+	}
+
+	err := options.Validate()
+	if err != nil {
+		return nil, len(workItems), 0, fmt.Errorf("Failed to validate job options: '%v'.", err)
+	}
+
+	completeItems := make([]OutputType, 0)
+	remainingItems := workItems
+
+	// If we are not overwriting the cache, query for cached results.
+	if !options.OverwriteCache && cacheFunc != nil {
+		completeItems, remainingItems, err = cacheFunc(workItems)
+		if err != nil {
+			return nil, len(workItems), 0, err
+		}
+	}
+
+	runTime := int64(0)
+	results := make([]OutputType, 0)
+
+	if options.WaitForCompletion {
+		results, runTime, err = runJob(options, remainingItems, cacheFunc, removeCacheFunc, workFunc)
+		if err != nil {
+			return nil, len(workItems), 0, err
+		}
+
+		// If the context was canceled during execution, return immediately.
+		if options.Context.Err() != nil {
+			return nil, 0, 0, nil
+		}
+
+		completeItems = append(completeItems, results...)
+		remainingItems = nil
+	} else {
+		go func() {
+			_, _, err := runJob(options, remainingItems, cacheFunc, removeCacheFunc, workFunc)
+			if err != nil {
+				log.Error("Failure while running asynchronous job.")
+			}
```

Comment:
If we have an output struct (whether or not that is apart of the input struct),
then we could do something pretty cool here.
We can have a receive-only chan in the struct that tells the caller when the job is actually done (the chan will be closed on done, like a context).
On WaitForCompletion, that chan will be closed before we return, but for async we can close in this gofunc after we actually finish.